### PR TITLE
#59: Added design document

### LINKED ISSUE #59: Prepare design document
- Put  general overview how the extension works

### PR DESCRIPTION
**Pull request by umitbuyuksahin**
Fixes #59

### CODE REVIEW

File: doc/design.md [hunk 1]
```diff
@@ -0,0 +1,31 @@
+# Design
+
+## Introduction
+The Transformers Extension proposes to use large pre-trained models using 
+[Transformers API](https://huggingface.co/docs/transformers/index) in Exasol 
+Database. These models can be used for many predictive tasks such as text 
+classification, question-answering, text generation  etc. in the field of NLP.
```
Comment (ckunki) [inline doc/design.md:7]:
```suggestion
The Transformers Extension proposes to use large pre-trained models together with Exasol Database's [Transformers API](https://huggingface.co/docs/transformers/index). These models can be used for many predictive tasks such as text 
classification, question-answering, text generation etc. in the field of NLP.
```

File: doc/design.md [hunk 2]
```diff
@@ -0,0 +1,31 @@
...
+
+
+## Design Considerations:
+Pre-trained models contain many parameters and can reach very large size. 
+Therefore, it is not efficient approach to download these large models in 
+each predictive task. In order to overcome this challenge, we proposed to 
+divide the functionality of this extension into two different parts:
```
Comment (ckunki) [inline doc/design.md:14]:
```suggestion
Pre-trained models contain many parameters and can grow very large.  Therefore, it is not efficient to download these large models in each predictive task. Transformer Extensions is therefore divided into two different parts:
```

File: doc/design.md [hunk 3]
```diff
@@ -0,0 +1,31 @@
...
+
+1. Download : It is responsible to download the specified pre-defined model 
+into the Exasol BucketFS.
+2. Prediction: These are a group of UDFs for each supported task. Each of 
+them caches the downloaded pre-trained model and perform prediction.
```
Comment (ckunki) [inline doc/design.md:19]:
```suggestion
1. Download: Responsible to download the specified pre-defined model into the Exasol BucketFS.
2. Prediction: Contains a UDF for each supported task. Each UDF caching the downloaded pre-trained model and performing prediction. 
```

Comment (ckunki) [reply doc/design.md:19]: Is this correct? Or can it be that a single task is made up of more than one UDF?

File: doc/design.md [hunk 4]
```diff
@@ -0,0 +1,31 @@
...
+Pre-trained models contain many parameters and can reach very large size. 
+Therefore, it is not efficient approach to download these large models in 
+each predictive task. In order to overcome this challenge, we proposed to 
+divide the functionality of this extension into two different parts:
+
+1. Download : It is responsible to download the specified pre-defined model 
+into the Exasol BucketFS.
+2. Prediction: These are a group of UDFs for each supported task. Each of 
+them caches the downloaded pre-trained model and perform prediction. 
+These supported tasks:
```
Comment (ckunki) [inline doc/design.md:20]:
```suggestion
Supported tasks:
```

File: doc/design.md [hunk 5]
```diff
@@ -0,0 +1,31 @@
...
+   1. Sequence Classification for Single Text
+   2. Sequence Classification for Text Pair
+   3. Question Answering
+   4. Masked Language Modelling
```
Comment (ckunki) [inline doc/design.md:24]:
```suggestion
   4. Masked Language Modeling
```

**Review by ckunki — CHANGES_REQUESTED** [review]

File: doc/design.md (hunk 1 above)
Comment (umitbuyuksahin) [reply doc/design.md:7]: Transformers API is not the one we provide. This is the one we use.

File: doc/design.md (hunk 3 above)
Comment (umitbuyuksahin) [reply doc/design.md:16]: yes, it is correct, thanks.

**Review by ckunki — APPROVED** [review]

**Pull request merged**